### PR TITLE
uc-crux-llvm: Skip execution of declared (not defined) functions

### PR DESCRIPTION
**Pull request by langston-barrett**
The goal here is more coverage, at the expense of soundness/safety guarantees. Unsoundness is tracked and reported to the user.

### CODE REVIEW

File: crucible/src/Lang/Crucible/FunctionHandle.hs [hunk 1]
```diff
@@ -231,3 +232,7 @@ searchHandleMap nm fnTyRepr (FnHandleMap m) =
       case testEquality (handleType h) fnTyRepr of
         Just Refl -> Just (h,x)
         Nothing -> Nothing
+
+handleMapToHandles :: FnHandleMap f -> [SomeHandle]
+handleMapToHandles (FnHandleMap m) =
+  map (\(Some (HandleElt handle _)) -> SomeHandle handle) (MapF.elems m)
```
Comment (langston-barrett) [inline crucible/src/Lang/Crucible/FunctionHandle.hs:238]: @robdockins Is this alright? Would this break an important abstraction barrier?

I'm using this [here](https://github.com/GaloisInc/crucible/pull/703/files#diff-e76f47d5b800a8b1d75a4bed25c18ba59ed5181cc7b4538079604b32c7f063f1R89-R97), the point is that I want to define no-op overrides for functions that have declarations but not definitions, but I need to *not* do that for functions that already have corresponding CFGs, like `free`.

File: uc-crux-llvm/test/Test.hs [hunk 2]
```diff
@@ -357,11 +363,12 @@ inFileTests =
...
         ("branch.c", [("branch", isSafe mempty)]),
         ("compare_to_null.c", [("compare_to_null", isSafe mempty)]),
+        ("extern_void_function.c", [("extern_void_function", isSafe (skipOverride "do_stuff"))]),
         -- This override needs refinement; the following should be safe with the
         -- precondition that the argument pointer is valid.
-        ("getenv_arg.c", [("getenv_arg", isSafe (usesOverride "getenv"))]),
-        ("getenv_const.c", [("getenv_const", isSafe (usesOverride "getenv"))]),
-        ("gethostname_const_len.c", [("gethostname_const_len", isSafe (usesOverride "gethostname"))]),
+        ("getenv_arg.c", [("getenv_arg", isSafe (unsoundOverride "getenv"))]),
+        ("getenv_const.c", [("getenv_const", isSafe (unsoundOverride "getenv"))]),
```
Comment (langston-barrett) [inline uc-crux-llvm/test/Test.hs:370]: The code here is mostly complete (insofar as the context of this PR, there are TODOs around the extension to non-void functions that I want to tackle soon), but these two tests have started failing for reasons I don't yet understand.
```
    getenv_arg.c:                                                      FAIL (0.06s)
      test/Test.hs:202:
      expected: Unsoundness {unsoundOverridesUsed = fromList [UnsoundOverrideName {getUnsoundOverrideName = "getenv"}], unsoundSkipOverridesUsed = fromList []}
       but got: Unsoundness {unsoundOverridesUsed = fromList [], unsoundSkipOverridesUsed = fromList []}
    getenv_const.c:                                                    FAIL (0.06s)
      test/Test.hs:202:
      expected: Unsoundness {unsoundOverridesUsed = fromList [UnsoundOverrideName {getUnsoundOverrideName = "getenv"}], unsoundSkipOverridesUsed = fromList []}
       but got: Unsoundness {unsoundOverridesUsed = fromList [], unsoundSkipOverridesUsed = fromList []}
```
The strange thing here is that I didn't really change anything about how the unsound overries operate...

Comment (langston-barrett) [reply uc-crux-llvm/test/Test.hs:370]: If I remove the call to `registerUnsoundSkipOverrides`, then these tests pass.

Comment (langston-barrett) [reply uc-crux-llvm/test/Test.hs:370]: It's not that a different override is getting registered for `getenv`... With this diff:
```diff
diff --git a/uc-crux-llvm/src/UCCrux/LLVM/Overrides/Skip.hs b/uc-crux-llvm/src/UCCrux/LLVM/Overrides/Skip.hs
index 9eb03c9d..05e3a5a6 100644
--- a/uc-crux-llvm/src/UCCrux/LLVM/Overrides/Skip.hs
+++ b/uc-crux-llvm/src/UCCrux/LLVM/Overrides/Skip.hs
@@ -12,6 +12,7 @@ Stability    : provisional
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TupleSections #-}
 
 module UCCrux.LLVM.Overrides.Skip
   ( SkipOverrideName (..),
@@ -93,9 +94,12 @@ registerUnsoundSkipOverrides proxy llvmModule mtrans usedRef decls =
               (handleMapToHandles (fnBindings binds))
     let overrides =
           mapMaybe
-            (createSkipOverride proxy usedRef)
+            (\dec -> (dec,) <$>  createSkipOverride proxy usedRef dec)
             (filter ((`Set.notMember` alreadyDefined) . declName) decls)
-    register_llvm_overrides llvmModule [] overrides llvmCtx
+    mapM_
+      (\dec -> liftIO (putStrLn ("Skipping " ++ Text.unpack (declName dec))))
+      (map fst overrides)
+    register_llvm_overrides llvmModule [] (map snd overrides) llvmCtx
```
I only see `llvm.dbg.value` getting skipped.

Maybe it's that I'm calling `register_llvm_overrides` twice? I'll look deeper into that function.

Comment (langston-barrett) [reply uc-crux-llvm/test/Test.hs:370]: Yep, `register_llvm_overrides` can only be called once. Fixed!

File: crucible/src/Lang/Crucible/FunctionHandle.hs (hunk 1 above)
Comment (travitch) [reply crucible/src/Lang/Crucible/FunctionHandle.hs:238]: It doesn't look like the `FnHandleMap` maintains any invariants

**Review by travitch — APPROVED** [review]

**Comment by langston-barrett** [comment]
The CI failure is the cache corruption one again:
```
[ 2 of 51] Compiling Lang.Crucible.Panic ( src/Lang/Crucible/Panic.hs, /Users/runner/work/crucible/crucible/dist-newstyle/build/x86_64-osx/ghc-8.8.4/crucible-0.5/build/Lang/Crucible/Panic.o )
<command line>: dlopen(/Users/runner/work/crucible/crucible/dist-newstyle/build/x86_64-osx/ghc-8.8.4/what4-1.1.0.0.99/build/libHSwhat4-1.1.0.0.99-inplace-ghc8.8.4.dylib, 5): no suitable image found.  Did find:
	/Users/runner/work/crucible/crucible/dist-newstyle/build/x86_64-osx/ghc-8.8.4/what4-1.1.0.0.99/build/libHSwhat4-1.1.0.0.99-inplace-ghc8.8.4.dylib: unknown file type, first eight bytes: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
	/Users/runner/work/crucible/crucible/dist-newstyle/build/x86_64-osx/ghc-8.8.4/what4-1.1.0.0.99/build/libHSwhat4-1.1.0.0.99-inplace-ghc8.8.4.dylib: unknown file type, first eight bytes: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
cabal: Failed to build crucible-0.5 (which is required by exe:crux-mir from
crux-mir-0.4).
```
There's no way that this PR affects `crux-mir`, though so as long as the failures are there that seems OK.

**Pull request merged**